third_party/pcsc-lite/naclport/server/src/readerfactory_nacl.cc

```diff
@@ -0,0 +1,110 @@
+// INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+// NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+// DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+// THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+// (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+// THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+
+// This file contains a replacement function for the original readerfactory.c
+// PC/SC-Lite internal implementation.
+
+// TODO: These includes need to come before readerfactory.h otherwise compiler
+//       goes nuts. Why is that?
+#include <ppapi/cpp/var.h>
+#include <ppapi/cpp/var_dictionary.h>
+#include <ppapi/cpp/instance.h>
+#include <ppapi/cpp/module.h>
+#include <google_smart_card_common/pp_var_utils/construction.h>
+
+extern "C" {
+#include "readerfactory.h"
+
+LONG RFAddReaderServer(const char *readerNameLong, int port,
+    const char *library, const char *device);
+LONG RFRemoveReaderServer(const char *readerName, int port);
+}
+
+#include <google_smart_card_common/logging/syslog/syslog.h>
+
+using namespace google_smart_card;
+
+const char kTypeMessageKey[] = "type";
+const char kDataMessageKey[] = "data";
+const char kAddReaderMessageType[] = "reader_add";
+const char kRemoveReaderMessageType[] = "reader_remove";
+const char kNameMessageKey[] = "readerName";
+const char kPortMessageKey[] = "port";
+const char kDeviceMessageKey[] = "device";
+const char kReturnCodeMessageKey[] = "returnCode";
+
+void post_message(const char* type, const pp::VarDictionary& message_data)
+{
+  pp::VarDictionary message;
+  message.Set(kTypeMessageKey, type);
+  message.Set(kDataMessageKey, message_data);
+
+  const pp::Module* const pp_module = pp::Module::Get();
+  if (pp_module) {
+    const pp::Module::InstanceMap pp_instance_map =
+        pp_module->current_instances();
+    for (const auto& instance_map_item : pp_instance_map) {
+      pp::Instance* const instance = instance_map_item.second;
+
+      if (instance) instance->PostMessage(message);
+    }
+  }
+}
+
+LONG RFAddReader(const char *readerNameLong, int port, const char *library,
+    const char *device)
+{
+  post_message(kAddReaderMessageType, VarDictBuilder()
+      .Add(kNameMessageKey, readerNameLong).Add(kPortMessageKey, port)
+      .Add(kDeviceMessageKey, device).Result());
+
+  LONG ret = RFAddReaderServer(readerNameLong, port, library, device);
+
+  // TODO: Testing code, to be removed at a later time (failing reader).
+  if (std::string("SCM Microsystems Inc. SCR 3310") == readerNameLong) {
+    post_message(kAddReaderMessageType, VarDictBuilder()
+        .Add(kNameMessageKey, readerNameLong).Add(kPortMessageKey, port)
+        .Add(kDeviceMessageKey, device).Add(kReturnCodeMessageKey, SCARD_E_INVALID_VALUE).Result());
+    return ret;
+  }
+
+  post_message(kAddReaderMessageType, VarDictBuilder()
```

Comment:
I would say that it's cleaner to use here a different message type than in the real "add" even that happens in the line 79. Here it's actually an "update" of the reader.